### PR TITLE
Refactor purge-service-offering command

### PR DESCRIPTION
**Pull request by AartiKriplani**
This PR refactors the `purge-service-offering` command.

More details can be found [here](https://www.pivotaltracker.com/story/show/161378850).
 
Note on providing the -p flag against an old environment:
There is no official capi-release that extends this far back. It will be extremely hard to reproduce.

We can see back in CC 2.46.0 that the /v2/services endpoint has provider filter (http://apidocs.cloudfoundry.org/227/services/list_all_services.html)
In CC 2.51.0 this filter is no longer possible (http://apidocs.cloudfoundry.org/231/services/list_all_services.html)

Nevertheless, we have added the logic to filter on this to the best of our knowledge:

https://github.com/cloudfoundry-incubator/cli-sapi/blob/purge-service-offering-refactor/command/v6/purge_service_offering_command.go#L68-L72
https://github.com/cloudfoundry-incubator/cli-sapi/blob/purge-service-offering-refactor/actor/v2action/service.go#L63-L67

Aarti,
On Behalf of SAPI team

### CODE REVIEW

File: actor/v2action/service.go [hunk 1]
```diff
@@ -52,6 +52,31 @@ func (actor Actor) GetServiceByNameAndBrokerName(serviceName, serviceBrokerName
 	return Service(services[0]), Warnings(warnings), nil
 }
 
+//TODO: look at common code between GetServiceByNameandProvider and GetServiceByName
```
Comment (williammartin) [inline actor/v2action/service.go]: We need to remove this TODO and add Go doc

File: actor/v2action/service.go [hunk 2]
```diff
@@ -8,13 +8,19 @@ import (
 
 type Service ccv2.Service
 
+// ServicesWithPlans is an association between a Service and the plans it offers.
+type ServicesWithPlans map[Service][]ServicePlan
+
+type Filter ccv2.Filter
```
Comment (tjvman) [inline actor/v2action/service.go]: Is this being used anywhere?

File: api/cloudcontroller/ccv2/service_test.go [hunk 3]
```diff
@@ -489,4 +489,71 @@ var _ = Describe("Service", func() {
 			})
 		})
 	})
+
+	Describe("DeleteService", func() {
```
[the remark this answers is not in the document]
Comment (williammartin) [reply api/cloudcontroller/ccv2/service_test.go]: Done

File: command/v6/purge_service_offering_command.go [hunk 4]
```diff
@@ -1,23 +1,131 @@
...
 	RequiredArgs    flag.Service `positional-args:"yes"`
 	Force           bool         `short:"f" description:"Force deletion without confirmation"`
 	Provider        string       `short:"p" description:"Provider"`
 	usage           interface{}  `usage:"CF_NAME purge-service-offering SERVICE [-p PROVIDER] [-f]\n\nWARNING: This operation assumes that the service broker responsible for this service offering is no longer available, and all service instances have been deleted, leaving orphan records in Cloud Foundry's database. All knowledge of the service will be removed from Cloud Foundry, including service instances and service bindings. No attempt will be made to contact the service broker; running this command without destroying the service broker will cause orphan service instances. After running this command you may want to run either delete-service-auth-token or delete-service-broker to complete the cleanup."`
 	relatedCommands interface{}  `related_commands:"marketplace, purge-service-instance, service-brokers"`
+
+	UI          command.UI
+	SharedActor command.SharedActor
+	Actor       PurgeServiceOfferingActor
+	otherActor  ServiceActor
```
[the remark this answers is not in the document]
Comment (williammartin) [reply command/v6/purge_service_offering_command.go]: No. I don't know how that got there.

File: integration/shared/isolated/purge_service_offering_command_test.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+						Eventually(session).Should(Say("WARNING: This operation assumes that the service broker responsible for this service offering is no longer available, and all service instances have been deleted, leaving orphan records in Cloud Foundry's database\\. All knowledge of the service will be removed from Cloud Foundry, including service instances and service bindings\\. No attempt will be made to contact the service broker; running this command without destroying the service broker will cause orphan service instances\\. After running this command you may want to run either delete-service-auth-token or delete-service-broker to complete the cleanup\\."))
+						Eventually(session).Should(Say("Really purge service offering %s from Cloud Foundry?", service))
+						Eventually(session).ShouldNot(Say("Purging service %s...", service))
+						Eventually(session).ShouldNot(Say("OK"))
+						Eventually(session).Should(Exit(0))
+					})
+				})
+
+				When("the -f flag is provided", func() {
+					It("purges the service offering without asking for confirmation", func() {
```
[the remark this answers is not in the document]
Comment (williammartin) [reply integration/shared/isolated/purge_service_offering_command_test.go]: Done thanks

File: integration/helpers/service_broker.go [hunk 6]
```diff
@@ -153,7 +153,8 @@ func (b ServiceBroker) Delete() {
 }
 
 func (b ServiceBroker) Destroy() {
-	Eventually(CF("purge-service-offering", b.Service.Name, "-f")).Should(Exit(0))
+	// We ignore the exit code for purge service offering, since it is not idempotent.
+	Eventually(CF("purge-service-offering", b.Service.Name, "-f")).Should(Exit())
```
[the remark this answers is not in the document]
Comment (williammartin) [reply integration/helpers/service_broker.go]: We decided to revert this behaviour to match the unrefactored behaviour and be idempotent.

File: actor/v2action/service.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (williammartin) [reply actor/v2action/service.go]: Yes it got merged in with `service-access` refactor and is used in `GetServicesWithPlans`. I don't think it belongs in this file though. I'll open a follow up to define it in a more useful place.

File: command/v6/purge_service_offering_command.go [hunk 7]
```diff
@@ -1,23 +1,131 @@
...
+		service, warnings, err = cmd.Actor.GetServiceByNameAndProvider(cmd.RequiredArgs.Service, cmd.Provider)
+	} else {
+		service, warnings, err = cmd.Actor.GetServiceByNameAndBrokerName(cmd.RequiredArgs.Service, "")
+	}
+
+	if err != nil {
+		cmd.UI.DisplayWarnings(warnings)
+
+		if _, ok := err.(actionerror.ServiceNotFoundError); ok {
+			return actionerror.PurgeServiceOfferingNotFoundError{Name: cmd.RequiredArgs.Service}
```
[the remark this answers is not in the document]
Comment (williammartin) [reply command/v6/purge_service_offering_command.go]: Agree, I think we are going to drop support for `-p` though based on your notes above.

**Comment by tjvman** [comment]
It looks like there's still a bit of code hanging around for checking max version even though `-p` will error well before that version check occurs. Seems like that code could be removed.

**Comment by williammartin** [comment]
This is currently failing in our CI so no mergy.

**Comment by tjvman** [comment]
After this change, the `purge-service-offering` does not support `-b`, and the refactored code [returns an error if it can't unambiguously locate a specific service](https://github.com/cloudfoundry/cli/blob/b82c7c3e274c4817885b011f87fa7b40ad657fb7/actor/v2action/service.go#L50). But there are [test cleanups](https://github.com/cloudfoundry/cli/blob/b82c7c3e274c4817885b011f87fa7b40ad657fb7/integration/shared/isolated/enable_service_access_command_test.go#L222) that are relying on the fact that the legacy implementation of `purge-service-offering` would pick a service in the case when there were multiple services with the same name.

Can you change this test clean up so that it works, and add a TODO to change it back to using the simple helper method after `purge-service-offering` starts accepting `-b`?

**Comment by nmaslarski** [comment]
Hay @tjvman,
We decided to make the command experimental, and execute the old command, until we implement the `-b` flag.
I pushed it in a WIP branch since I don't have the time to wait the CI
https://github.com/cloudfoundry-incubator/cli-sapi/tree/wip-purge-service-offering-refactor
I'll include it in the PR tomorrow morning if everything is fine.

**Comment by henryaj** [comment]
Waiting for CI to go through before we merge.

**Pull request merged**